File: packages/core/src/math/routing.ts
/**
 * Deterministic multi-asset route selection.
 *
 * The selector consumes only normalized integer quotes. It performs no I/O,
 * reads no clock, uses no floating point, and defines a total tie-break order.
 */
import type { RouteQuote } from '../routing/types.js';

export const ROUTING_WEIGHT_SCALE = 10_000;

export interface RoutingPolicy {
  /** Relative weight of source-normalized fees. */
  costWeight: number;
  /** Relative weight of expected price impact. */
  slippageWeight: number;
  /** Relative weight of the reliability shortfall. */
  reliabilityWeight: number;
  /** Fixed score added for every economic hop after the first. */
  hopPenalty: number;
  /** Routes above this expected slippage are inadmissible. */
  maxSlippageBps: number;
  /** Routes below this reliability are inadmissible. */
  minReliabilityBps: number;
}

export const DEFAULT_ROUTING_POLICY: Readonly<RoutingPolicy> = Object.freeze({
  costWeight: 5_000,
  slippageWeight: 3_000,
  reliabilityWeight: 2_000,
  hopPenalty: 5,
  maxSlippageBps: 1_000,
  minReliabilityBps: 5_000,
});

export interface RouteScoreBreakdown {
  weightedCost: string;
  weightedSlippage: string;
  weightedReliability: string;
  hopPenalty: string;
}

export interface ScoredRoute extends RouteQuote {
  /** Lower is better. Integer score for byte-identical TS/Python output. */
  score: string;
  breakdown: RouteScoreBreakdown;
}

/** Validate and score one normalized route, including policy admission bounds. */
export function scoreRoute(
  route: RouteQuote,
  policy: RoutingPolicy = DEFAULT_ROUTING_POLICY,
): ScoredRoute {
  validatePolicy(policy);
  validateRoute(route);
  if (!isRouteEligible(route, policy)) {
    throw new RangeError(`route ${route.id} is outside routing policy bounds`);
  }

  const weightedCost = weighted(route.totalFeeBps, policy.costWeight);
  const weightedSlippage = weighted(route.expectedSlippageBps, policy.slippageWeight);
  const reliabilityShortfall = ROUTING_WEIGHT_SCALE - route.reliabilityBps;
  const weightedReliability = weighted(reliabilityShortfall, policy.reliabilityWeight);
  const hopPenalty = BigInt(route.hopCount - 1) * BigInt(policy.hopPenalty);
  const score = weightedCost + weightedSlippage + weightedReliability + hopPenalty;

  return {
    ...route,
    score: score.toString(),
    breakdown: {
      weightedCost: weightedCost.toString(),
      weightedSlippage: weightedSlippage.toString(),
      weightedReliability: weightedReliability.toString(),
      hopPenalty: hopPenalty.toString(),
    },
  };
}

/** Whether a structurally valid route clears policy reliability/slippage bounds. */
export function isRouteEligible(
  route: RouteQuote,
  policy: RoutingPolicy = DEFAULT_ROUTING_POLICY,
): boolean {
  return route.expectedSlippageBps <= policy.maxSlippageBps &&
    route.reliabilityBps >= policy.minReliabilityBps;
}

/**
 * Rank routes with a total, input-order-independent comparison:
 * score, output (descending), slippage, hop count, then canonical route ID.
 */
export function rankRoutes(
  routes: readonly RouteQuote[],
  policy: RoutingPolicy = DEFAULT_ROUTING_POLICY,
): ScoredRoute[] {
  validatePolicy(policy);
  const scored = routes
    .filter((route) => {
      validateRoute(route);
      return isRouteEligible(route, policy);
    })
    .map((route) => scoreRoute(route, policy));
  scored.sort(compareScoredRoutes);
  return scored;
}

/** Select the deterministic winner, or null when no route is admissible. */
export function selectRoute(
  routes: readonly RouteQuote[],
  policy: RoutingPolicy = DEFAULT_ROUTING_POLICY,
): ScoredRoute | null {
  return rankRoutes(routes, policy)[0] ?? null;
}

export function validateRoutingPolicy(policy: RoutingPolicy): void {
  validatePolicy(policy);
}

function compareScoredRoutes(a: ScoredRoute, b: ScoredRoute): number {
  const scoreA = BigInt(a.score);
  const scoreB = BigInt(b.score);
  if (scoreA !== scoreB) return scoreA < scoreB ? -1 : 1;
  const outputA = BigInt(a.expectedDestinationAmount);
  const outputB = BigInt(b.expectedDestinationAmount);
  if (outputA !== outputB) return outputA > outputB ? -1 : 1;
  if (a.expectedSlippageBps !== b.expectedSlippageBps) {
    return a.expectedSlippageBps - b.expectedSlippageBps;
  }
  if (a.hopCount !== b.hopCount) return a.hopCount - b.hopCount;
  return binaryCompare(a.id, b.id);
}

function weighted(value: number, weight: number): bigint {
  return (BigInt(value) * BigInt(weight)) / BigInt(ROUTING_WEIGHT_SCALE);
}

function validatePolicy(policy: RoutingPolicy): void {
  for (const [name, value] of Object.entries(policy)) {
    if (!Number.isSafeInteger(value) || value < 0) {
      throw new RangeError(`${name} must be a non-negative safe integer`);
    }
  }
  const weightSum = policy.costWeight + policy.slippageWeight + policy.reliabilityWeight;
  if (weightSum !== ROUTING_WEIGHT_SCALE) {
    throw new RangeError(`routing weights must sum to ${ROUTING_WEIGHT_SCALE}, got ${weightSum}`);
  }
  if (policy.maxSlippageBps > ROUTING_WEIGHT_SCALE ||
    policy.minReliabilityBps > ROUTING_WEIGHT_SCALE) {
    throw new RangeError('routing basis-point bounds must not exceed 10000');
  }
}

function validateRoute(route: RouteQuote): void {
  if (!route.id) throw new RangeError('route id is required');
  if (!/^[1-9][0-9]*$/.test(route.sourceAmount) ||
    !/^[1-9][0-9]*$/.test(route.expectedDestinationAmount)) {
    throw new RangeError('route amounts must be positive canonical integers');
  }
  for (const [name, value] of [
    ['totalFeeBps', route.totalFeeBps],
    ['expectedSlippageBps', route.expectedSlippageBps],
    ['reliabilityBps', route.reliabilityBps],
  ] as const) {
    if (!Number.isInteger(value) || value < 0 || value > ROUTING_WEIGHT_SCALE) {
      throw new RangeError(`${name} must be an integer from 0 to 10000`);
    }
  }
  if (!Number.isInteger(route.hopCount) || route.hopCount < 1) {
    throw new RangeError('hopCount must be a positive integer');
  }
}

function binaryCompare(a: string, b: string): number {
  if (a === b) return 0;
  const left = new TextEncoder().encode(a);
  const right = new TextEncoder().encode(b);
  const length = Math.min(left.length, right.length);
  for (let index = 0; index < length; index += 1) {
    if (left[index] !== right[index]) return left[index]! < right[index]! ? -1 : 1;
  }
  return left.length < right.length ? -1 : 1;
}

File: python/src/stellaragent/routing.py
"""Deterministic multi-asset route selection.

Strict Python port of ``packages/core/src/math/routing.ts``. Route amounts are
integer base-unit strings and all scoring uses integer arithmetic. Shared
fixtures prove that TypeScript and Python return identical scores and ordering.
"""

from __future__ import annotations

import re
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Any

__all__ = [
    "ROUTING_WEIGHT_SCALE",
    "RoutingPolicy",
    "DEFAULT_ROUTING_POLICY",
    "RouteScoreBreakdown",
    "ScoredRoute",
    "score_route",
    "is_route_eligible",
    "rank_routes",
    "select_route",
    "validate_routing_policy",
]

ROUTING_WEIGHT_SCALE = 10_000


@dataclass(frozen=True)
class RoutingPolicy:
    """Integer weights and admission bounds for route scoring."""

    cost_weight: int = 5_000
    slippage_weight: int = 3_000
    reliability_weight: int = 2_000
    hop_penalty: int = 5
    max_slippage_bps: int = 1_000
    min_reliability_bps: int = 5_000


DEFAULT_ROUTING_POLICY = RoutingPolicy()


@dataclass(frozen=True)
class RouteScoreBreakdown:
    weighted_cost: str
    weighted_slippage: str
    weighted_reliability: str
    hop_penalty: str


@dataclass(frozen=True)
class ScoredRoute:
    """A route plus its deterministic integer score."""

    route: Mapping[str, Any]
    score: str
    breakdown: RouteScoreBreakdown

    @property
    def id(self) -> str:
        return str(self.route["id"])


def score_route(
    route: Mapping[str, Any],
    policy: RoutingPolicy = DEFAULT_ROUTING_POLICY,
) -> ScoredRoute:
    """Validate and score one admitted normalized route."""
    validate_routing_policy(policy)
    _validate_route(route)
    if not is_route_eligible(route, policy):
        raise ValueError(f"route {route['id']} is outside routing policy bounds")

    weighted_cost = _weighted(_field_int(route, "totalFeeBps"), policy.cost_weight)
    weighted_slippage = _weighted(
        _field_int(route, "expectedSlippageBps"), policy.slippage_weight
    )
    shortfall = ROUTING_WEIGHT_SCALE - _field_int(route, "reliabilityBps")
    weighted_reliability = _weighted(shortfall, policy.reliability_weight)
    hop_penalty = (_field_int(route, "hopCount") - 1) * policy.hop_penalty
    score = weighted_cost + weighted_slippage + weighted_reliability + hop_penalty

    return ScoredRoute(
        route=dict(route),
        score=str(score),
        breakdown=RouteScoreBreakdown(
            weighted_cost=str(weighted_cost),
            weighted_slippage=str(weighted_slippage),
            weighted_reliability=str(weighted_reliability),
            hop_penalty=str(hop_penalty),
        ),
    )


def is_route_eligible(
    route: Mapping[str, Any],
    policy: RoutingPolicy = DEFAULT_ROUTING_POLICY,
) -> bool:
    return (
        _field_int(route, "expectedSlippageBps") <= policy.max_slippage_bps
        and _field_int(route, "reliabilityBps") >= policy.min_reliability_bps
    )


def rank_routes(
    routes: Sequence[Mapping[str, Any]],
    policy: RoutingPolicy = DEFAULT_ROUTING_POLICY,
) -> list[ScoredRoute]:
    """Return the same total ordering as the TypeScript selector."""
    validate_routing_policy(policy)
    admitted: list[ScoredRoute] = []
    for route in routes:
        _validate_route(route)
        if is_route_eligible(route, policy):
            admitted.append(score_route(route, policy))
    admitted.sort(
        key=lambda scored: (
            int(scored.score),
            -int(str(scored.route["expectedDestinationAmount"])),
            _field_int(scored.route, "expectedSlippageBps"),
            _field_int(scored.route, "hopCount"),
            scored.id.encode("utf-8"),
        )
    )
    return admitted


def select_route(
    routes: Sequence[Mapping[str, Any]],
    policy: RoutingPolicy = DEFAULT_ROUTING_POLICY,
) -> ScoredRoute | None:
    ranked = rank_routes(routes, policy)
    return ranked[0] if ranked else None


def validate_routing_policy(policy: RoutingPolicy) -> None:
    values = {
        "cost_weight": policy.cost_weight,
        "slippage_weight": policy.slippage_weight,
        "reliability_weight": policy.reliability_weight,
        "hop_penalty": policy.hop_penalty,
        "max_slippage_bps": policy.max_slippage_bps,
        "min_reliability_bps": policy.min_reliability_bps,
    }
    for name, value in values.items():
        if isinstance(value, bool) or not isinstance(value, int) or value < 0:
            raise ValueError(f"{name} must be a non-negative integer")
    weight_sum = policy.cost_weight + policy.slippage_weight + policy.reliability_weight
    if weight_sum != ROUTING_WEIGHT_SCALE:
        raise ValueError(
            f"routing weights must sum to {ROUTING_WEIGHT_SCALE}, got {weight_sum}"
        )
    if (
        policy.max_slippage_bps > ROUTING_WEIGHT_SCALE
        or policy.min_reliability_bps > ROUTING_WEIGHT_SCALE
    ):
        raise ValueError("routing basis-point bounds must not exceed 10000")


def _weighted(value: int, weight: int) -> int:
    return value * weight // ROUTING_WEIGHT_SCALE


def _validate_route(route: Mapping[str, Any]) -> None:
    if not isinstance(route.get("id"), str) or not route["id"]:
        raise ValueError("route id is required")
    for name in ("sourceAmount", "expectedDestinationAmount"):
        value = route.get(name)
        if not isinstance(value, str) or not re.fullmatch(r"[1-9][0-9]*", value):
            raise ValueError("route amounts must be positive canonical integers")
    for name in ("totalFeeBps", "expectedSlippageBps", "reliabilityBps"):
        value = _field_int(route, name)
        if value < 0 or value > ROUTING_WEIGHT_SCALE:
            raise ValueError(f"{name} must be an integer from 0 to 10000")
    hop_count = _field_int(route, "hopCount")
    if hop_count < 1:
        raise ValueError("hopCount must be a positive integer")


def _field_int(route: Mapping[str, Any], name: str) -> int:
    value = route.get(name)
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValueError(f"{name} must be an integer")
    return value
